doc/ps_design.md

```diff
@@ -0,0 +1,382 @@
+}
+
+service PServer{
+    rpc push_model(InitModel) returns (google.protobuf.Empty) {}
+    rpc pull_model(Model) returns (Model) {}
+    rpc pull_embedding_vector(Tensor) returns (EmbeddingResponse) {}
+    rpc push_gradient(Model) returns (google.protobuf.Empty)
+}
+```
+
+The interfaces of PServer could be like this:
+
+
+```python
+class PServer(elasticdl_pb2_grpc.PServerServicer):
+    def __init__(self, kvstore, grad_queue, opt):
+        self.kvstore = KVStore()
+        self.grad_queue = GradientQueue()
+        self.opt = Optimizer(opt, self.kvstore, self.grad_queue)
+
+    def pull_model(self, request, _):
+        pass
+
+    def push_model(self, request, _):
+        pass
+
+    def push_gradient(self, request, _):
+        pass
+
+    # embedding param is handled lazily
+    def pull_embedding_vector(self, reques, _):
+        pass
+```
+
+### Checkpoint and Serving
+
+Master will send signal to pservers to make checkpoint. Each pserver will save parameters in its current KVStore to a distributed file system.
+
+Since a pserver only has a subset of the whole model, we have to merge these submodels to get final model for serving.
+
+
+## Interactions among Master, PS and Worker
+
+The following events involve interactions among the master, workers and PS:
+
+* The master starts PS.
+* Initialization of parameters in PS.
+* Relaunch of PS.
+* Workers get model parameters from PS.
+* Workers push gradients to PS.
+* PS reports submodel version to the master.
+* The master tells PS to save checkpoint.
+
+
+### The master starts PS
+When an ElasticDL task starts, `master.main` is responsible for starting PS. After starting PS, `master.main` starts the master and workers, and tells them the endpoints of PS.
+
+### Initialization of parameters in PS
+PS does not have any model variable and mdoel meta info after starting. Model meta info includes dimension of embedding layers, initialization methods of embedding vectors, initialization methods of slot variables in optimizer.
+
+There are two ways for PS to get model variables and model meta info, one is to read from a checkpoint file, one is to obtain them from workers.
+
+When `master.main` starts PS, `master.main` decides how to initialize PS. If `master.main` passes an argument specifying the checkpoint file name to PS, PS reads from the checkpoint. Otherwise, PS does nothing but waiting for the first `get_model` from worker. In the reponse of `get_model` call, PS tells the worker to initialize model, and report model variables and meta info to the PS.
+
+Please Note that the worker only initializes model variables. ElasticDL adopts lazy initialization for embedding vectors. Please refer to "[Workers get model parameters from PS](#Workers-get-model-parameters-from-PS)" section.
+
+### Relaunch of PS
+In case a PS pod fails, the master will try to relaunch one PS and it should recover model variables and embedding tables.
+
+For model variables, PS can recover from workers.
```

Comment:
Worker will call `get_model`, but find that the new PS has no model on it. So the new PS will request worker to send its model.

The logic here is the same with parameter Initialization of PS.

Reply:
Yes, that's the logic I want to say. I will modify this sentence to make things clear.